Fake quantization enhancements for QAT/PTQ support

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26627 Default observer and fake-quant for backends
* #26626 Moving average for statistics in observers
* #26625 Emulate weight and activation only quant with fake quant, numerics test
* #26624 Quantization aware training: Freeze batch norm support
* #26623 Per channel fake quant
* #26516 Per-channel baseline
* #26520 control of observer/fake-quant operations
* #26622 Default histogram observer
* #26519 Serialization and range reduction support for Fake Quant/Observer
* **#26420 Fake quantization enhancements for QAT/PTQ support**
* #26518 Handle DeQuantStub() for QAT

Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

raghuramank10000 committed:
Fake quantization enhancements for QAT/PTQ support

Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -25,31 +25,45 @@ def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
-            self.observer(X)
+        if self.observer_enabled:
+            X = self.observer(X)
             scale, zero_point = self.calculate_qparams()
```
`observer_enabled` also means update `self.scale` and `self.zero_point`?

raghuramank100 replied:
Yes, thats the intent. In a later PR we can optimize this further to have a seperate flag for controlling scale/zero-point calculation

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LG

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fake quantization enhancements for QAT/PTQ support"


Flags for enabling/disabling observer and fake quant independently. Improve repr for fake quant.

Differential Revision: [D17458232](https://our.internmc.facebook.com/intern/diff/D17458232/)

[ghstack-poisoned]
